core/src/services/vercel_blob/core.rs

```diff
@@ -0,0 +1,433 @@
+            percent_encode_path(url)
+        ));
+
+        let req = self.sign(req);
+
+        // Set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn copy(&self, from: &str, to: &str) -> Result<Response<IncomingAsyncBody>> {
+        let from = build_abs_path(&self.root, from);
+
+        let resp = self.list(&from, Some(1)).await?;
+
+        let blobs = resp.blobs;
+
+        if blobs.is_empty() {
+            return Err(Error::new(ErrorKind::NotFound, "Blob not found"));
+        }
+
+        let from_url = &blobs[0].url;
+
+        let to = build_abs_path(&self.root, to);
+
+        let to_url = format!(
+            "https://blob.vercel-storage.com/{}?fromUrl={}",
+            percent_encode_path(&to),
+            percent_encode_path(from_url),
+        );
+
+        let req = Request::put(&to_url);
+
+        let req = self.sign(req);
+
+        // Set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn list(&self, prefix: &str, limit: Option<usize>) -> Result<ListResponse> {
+        let prefix = if prefix == "/" { "" } else { prefix };
+
+        let mut url = format!(
+            "https://blob.vercel-storage.com?prefix={}",
+            percent_encode_path(prefix)
+        );
+
+        if let Some(limit) = limit {
+            url.push_str(&format!("&limit={}", limit))
+        }
+
+        let req = Request::get(&url);
+
+        let req = self.sign(req);
+
+        // Set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        let resp = self.send(req).await?;
+
+        let status = resp.status();
+
+        match status {
+            StatusCode::OK => {
+                let body = resp.into_body().bytes().await?;
+
+                let resp: ListResponse =
+                    serde_json::from_slice(&body).map_err(new_json_deserialize_error)?;
+
+                Ok(resp)
+            }
+            _ => Err(parse_error(resp).await?),
+        }
+    }
+
+    pub async fn initiate_multipart_upload(
+        &self,
+        path: &str,
+        args: &OpWrite,
+    ) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://blob.vercel-storage.com/mpu/{}",
+            percent_encode_path(&p)
+        );
+
+        let req = Request::post(&url);
+
+        let mut req = self.sign(req);
+
+        // https://github.com/vercel/storage/blob/main/packages/blob/src/put-multipart.ts#L130
+        // x-mpu-action specifies the action to perform on the MPU.
+        // In this case, we want to create a new MPU.
+        req = req.header("x-mpu-action", "create");
+
+        req = req.header("x-add-random-suffix", "0");
```

Comment:
These headers are used in multiple places. We can extract them as constants and add comments to explain their meaning and possible values.

Reply:
fixed